blockchain/indexers/addrindex_test.go

```diff
@@ -274,3 +283,448 @@ nextTest:
+	}
+	defer func() {
+		db.Close()
+		os.RemoveAll(dbPath)
+	}()
+
+	genesis := dcrutil.NewBlockDeepCopy(chaincfg.SimNetParams().GenesisBlock)
+	chain, err := newTChain(genesis)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	g, err := chaingen.MakeGenerator(chaincfg.SimNetParams())
+	if err != nil {
+		t.Fatalf("unexpected error %v", err)
+	}
+
+	addBlock := func(blockName string, firstBlock bool) *dcrutil.Block {
+		var blk *wire.MsgBlock
+		if firstBlock {
+			blk = g.CreateBlockOne(blockName, 0)
+		} else {
+			blk = g.NextBlock(blockName, nil, nil)
+			g.SaveTipCoinbaseOuts()
+		}
+
+		return dcrutil.NewBlockDeepCopy(blk)
+	}
+
+	// Add three blocks to the chain.
+	bk1 := addBlock("bk1", true)
+	err = chain.AddBlock(bk1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	bk2 := addBlock("bk2", false)
+	err = chain.AddBlock(bk2)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	bk3 := addBlock("bk3", false)
+	err = chain.AddBlock(bk3)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Initialize the tx index and address index. The tx index is required
+	// because the address index depends on it.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	subber := NewIndexSubscriber(ctx)
+	chainParams := chaincfg.SimNetParams()
+
+	_, err = NewTxIndex(subber, db, chain, chainParams)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	addrIdx, err := NewAddrIndex(subber, db, chain, chainParams)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Ensure the indexes got synced to bk3 on initialization.
+	txIdxTipHeight, txIdxTipHash, err := addrIdx.Tip()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if txIdxTipHeight != bk3.Height() {
+		t.Fatalf("expected tip height to be %d, got %d", bk3.Height(), txIdxTipHeight)
+	}
+
+	if *txIdxTipHash != *bk3.Hash() {
+		t.Fatalf("expected tip hash to be %s, got %s", bk3.Hash().String(),
+			txIdxTipHash.String())
+	}
+
+	addrIdxTipHeight, addrIdxTipHash, err := addrIdx.Tip()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if addrIdxTipHeight != bk3.Height() {
+		t.Fatalf("expected tip height to be %d, got %d", bk3.Height(), addrIdxTipHeight)
+	}
+
+	if *addrIdxTipHash != *bk3.Hash() {
+		t.Fatalf("expected tip hash to be %s, got %s", bk3.Hash().String(),
+			addrIdxTipHash.String())
+	}
+
+	// Ensure the address index remains in sync with the main chain when new
+	// blocks are connected.
+	bk4 := addBlock("bk4", false)
+	err = chain.AddBlock(bk4)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ntfn := &IndexNtfn{
+		NtfnType:          ConnectNtfn,
+		Block:             bk4,
+		Parent:            bk3,
+		PrevScripts:       nil,
+		IsTreasuryEnabled: false,
+	}
+
+	subber.Notify(ntfn)
+	time.Sleep(time.Millisecond * 150)
+
+	bk5 := addBlock("bk5", false)
+	err = chain.AddBlock(bk5)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ntfn = &IndexNtfn{
+		NtfnType:          ConnectNtfn,
+		Block:             bk5,
+		Parent:            bk4,
+		PrevScripts:       nil,
+		IsTreasuryEnabled: false,
+	}
+	subber.Notify(ntfn)
+	time.Sleep(time.Millisecond * 150)
```

Comment:
Is this sleep intending to wait long enough for the notification to be processed?  That would likely lead to flakey tests that intermittently fail when it takes longer than that amount of time.